### PR TITLE
feat(validator): humanized validation messages

### LINKED ISSUE #22: Humanize validator error messages

### PR DESCRIPTION
**Pull request by pmstss**
BREAKING CHANGE: The `verify` method has been reworked. In case of validation failure, the returned promise is resolved with an array of `ErrorObject` instances.

Before:
```ts
import { OpenAPIV2 } from '@har-sdk/types';
import { OASValidator, Validator } from '@har-sdk/validator';

const doc = {
  swagger: '2.0',
  host: 'http://localhost:3000',
  info: {
    title: 'Some valid API document',
    version: '1.0.0'
  },
  paths: {}
} as OpenAPIV2.Document;

const result = await new OASValidator().verify(doc);
console.log(result);
// {
//   valid: false,
//   errors: [
//     {
//       instancePath: '/info',
//       schemaPath: '#/required',
//       keyword: 'required',
//       params: { missingProperty: 'version' },
//       message: "must have required property 'version'"
//     }
//   ]
// }
```

After:

```ts
import { OpenAPIV2 } from '@har-sdk/types';
import { OASValidator, ErrorHumanizer } from '@har-sdk/validator';

const doc = {
  swagger: '2.0',
  host: 'localhost',
  info: {
    title: 'Some valid API document'
  },
  paths: {}
} as OpenAPIV2.Document;

const errors = await new OASValidator().verify(doc);
console.log(errors);
// [
//   {
//     instancePath: '/info',
//     schemaPath: '#/required',
//     keyword: 'required',
//     params: { missingProperty: 'version' },
//     message: "must have required property 'version'"
//   }
// ]
```
closes #22

### CODE REVIEW

File: packages/validator/src/validator/ErrorCondenser.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+//
+// 1. group all errors by path
+// 2. score them by message frequency
+// 3. select the most frequent messages (retaining all equally-frequent messages)
+// 4. concatenate the params of each occurrence of the most frequent message
+// 5. create one condensed error for the path
+
+export class ErrorCondenser {
+  // tree: instancePath -> message -> errors
+  private tree: Record<string, Record<string, ErrorObject[]>>;
```
Comment (derevnjuk) [inline packages/validator/src/validator/ErrorCondenser.ts:13]:
```suggestion
  private tree?: Record<string, Record<string, ErrorObject[]>>;
```

File: packages/validator/src/validator/ErrorCondenser.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        return obj;
+      },
+      { max: 0, messages: [] }
+    ).messages;
+  }
+
+  private mergeDuplicatedErrors(errors: ErrorObject[]): ErrorObject {
+    return errors.reduce(
+      (prev: ErrorObject, err: ErrorObject): ErrorObject =>
```
Comment (derevnjuk) [inline packages/validator/src/validator/ErrorCondenser.ts:82]:
```suggestion
      (prev: ErrorObject | undefined, err: ErrorObject): ErrorObject =>
```

File: packages/validator/src/validator/ErrorCondenser.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+export class ErrorCondenser {
+  // tree: instancePath -> message -> errors
+  private tree: Record<string, Record<string, ErrorObject[]>>;
+
+  constructor(private readonly errors: ReadonlyArray<ErrorObject>) {}
+
+  public condense(): ErrorObject[] {
+    if (!Array.isArray(this.errors)) {
+      return [];
+    }
```
Comment (derevnjuk) [inline packages/validator/src/validator/ErrorCondenser.ts:20]: This is a redundant check, you can remove it.

File: packages/validator/src/validator/ErrorCondenser.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    this.parseToTree();
```
Comment (derevnjuk) [inline packages/validator/src/validator/ErrorCondenser.ts:22]: Why don't call this method in the constructor? Wdyt?

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/src/schemas/postman/v2.1.0.json [hunk 5]
```diff
@@ -543,99 +535,118 @@
...
+                  "LOCK",
+                  "UNLOCK",
+                  "PROPFIND",
+                  "VIEW"
+                ]
+              },
+              {
+                "description": "The Custom HTTP method associated with this request.",
+                "type": "string"
+              }
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.1.0.json:588]: I don't think we should allow users to specify custom methods

File: packages/validator/src/schemas/postman/v2.1.0.json [hunk 6]
```diff
@@ -543,99 +535,118 @@
...
+            }
+          },
+          "proxy": {
+            "$ref": "#/definitions/proxy-config"
+          },
+          "certificate": {
+            "$ref": "#/definitions/certificate"
+          },
+          "method": {
+            "anyOf": [
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.1.0.json:563]: Should we rework it using `if/else` keywords?

File: packages/validator/src/schemas/postman/v2.1.0.json [hunk 7]
```diff
@@ -543,99 +535,118 @@
...
-                        "title": "UrlEncodedParameter",
+                    "required": ["key"]
+                  }
+                },
+                "formdata": {
+                  "type": "array",
+                  "items": {
+                    "type": "object",
+                    "title": "FormParameter",
+                    "anyOf": [
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.1.0.json:648]: Should we rework it using `if/else` keyworks?

File: packages/validator/src/schemas/postman/v2.1.0.json (hunk 6 above)
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.1.0.json:563]: Should we rework it using `if/else` keyworks?

File: packages/validator/src/schemas/postman/v2.0.0.json [hunk 8]
```diff
@@ -498,99 +489,118 @@
...
+                  "LOCK",
+                  "UNLOCK",
+                  "PROPFIND",
+                  "VIEW"
+                ]
+              },
+              {
+                "description": "The Custom HTTP method associated with this request.",
+                "type": "string"
+              }
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.0.0.json:542]: ditto

File: packages/validator/src/schemas/postman/v2.0.0.json [hunk 9]
```diff
@@ -498,99 +489,118 @@
...
-                        "title": "UrlEncodedParameter",
+                    "required": ["key"]
+                  }
+                },
+                "formdata": {
+                  "type": "array",
+                  "items": {
+                    "type": "object",
+                    "title": "FormParameter",
+                    "anyOf": [
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.0.0.json:602]: ditto

File: packages/validator/src/schemas/openapi/v3.0.0.json [hunk 10]
```diff
@@ -140,1078 +134,1506 @@
...
-      }
-    },
-    "anyOrExpression": {
-      "oneOf": [
+      },
+      "patternProperties": {
+        "^x-": {}
+      },
+      "additionalProperties": false,
+      "required": ["name", "in"],
```
Comment (derevnjuk) [inline packages/validator/src/schemas/openapi/v3.0.0.json:1128]: When `in: path`, `required` is required to be `true`. Is there any way to cover this case?

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/src/schemas/openapi/v2.0.0.json [hunk 11]
```diff
@@ -753,13 +782,14 @@
...
     "pathParameterSubSchema": {
+      "type": "object",
       "additionalProperties": false,
       "patternProperties": {
         "^x-": {
           "$ref": "#/definitions/vendorExtension"
         }
       },
-      "required": ["required"],
+      "required": ["name", "in", "type", "required"],
```
Comment (derevnjuk) [inline packages/validator/src/schemas/openapi/v2.0.0.json:792]: - By default only `in` and `name` fields are required. 
- `type` is required when `in` is not `body`. 
- `schema` must be required if `in` is `body`

Wdyt? Can we handle these constraints somehow?

File: packages/validator/src/schemas/postman/v2.0.0.json [hunk 12]
```diff
@@ -236,31 +235,26 @@
...
-            },
-            "type": {
-              "type": "string",
-              "description": "Holds the mime type of the raw description content. E.g: 'text/markdown' or 'text/html'.\nThe type is used to correctly render the description when generating documentation, or in the Postman app."
-            },
-            "version": {
-              "description": "Description can have versions associated with it, which should be put in this property."
-            }
+      "if": {
+        "type": ["string", "null"]
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.0.0.json:239]: Should we extract `null` under `then` keyword?

File: packages/validator/src/schemas/postman/v2.0.0.json [hunk 13]
```diff
@@ -740,30 +716,25 @@
...
-                  {
-                    "$ref": "#/definitions/header"
-                  },
-                  {
-                    "title": "Header",
-                    "type": "string"
-                  }
-                ]
+          "if": {
+            "type": ["string", "null"]
```
Comment (derevnjuk) [inline packages/validator/src/schemas/postman/v2.0.0.json:720]: ditto

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/tests/error-condenser.spec.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+describe('ErrorCondenser', () => {
+  const error: ErrorObject = {
+    keyword: 'enum',
+    instancePath: '/paths/~1path',
+    schemaPath: '#/properties/in/enum',
+    params: {},
+    message: 'msg'
+  };
+
+  it('should condense duplicated errors', () => {
```
Comment (derevnjuk) [inline packages/validator/tests/error-condenser.spec.ts:14]: Put all tests under a single test suite (e.g. `describe('condense')`).

File: packages/validator/tests/error-condenser.spec.ts [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+      {
+        ...error,
+        params: { dummy: new Array(100).fill(42) }
+      }
+    ]);
+  });
+
+  it('should handle empty input', () => {
+    new ErrorCondenser(null).condense().should.be.empty;
+  });
```
Comment (derevnjuk) [inline packages/validator/tests/error-condenser.spec.ts:35]:
```suggestion
  it('should handle empty input', () => new ErrorCondenser(null).condense().should.be.empty);
```

File: packages/validator/tests/error-humanizer.spec.ts [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+    openapi: '3.0.1',
+    servers: [{ url: 'localhost' }],
+    info: {
+      title: 'Invalid OpenAPI document',
+      version: '1.0.0'
+    },
+    paths: {}
+  });
+
+  it('should return original "errorMessage" if exists', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/error-humanizer.spec.ts:46]: ditto (e.g. `describe('humanizeErrors')`)

File: packages/validator/tests/error-humanizer.spec.ts [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+    const apiDoc: OpenAPIV2.Document = {
+      ...getBaseSwaggerDoc(),
+      paths: {
+        path1: {}
+      }
+    };
+
+    const result = humanizer
+      .humanizeErrors(await oasValidator.verify(apiDoc))
```
Comment (derevnjuk) [inline packages/validator/tests/error-humanizer.spec.ts:55]: To keep the test structure clean, I suggest using AAA pattern. Move `await oasValidator.verify(apiDoc)` under the arrange step.

```js
// arrange
const input = {};
const expected = {};

// act
const result = doSomething(input);

// assert
result.should.eq(input)
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/src/validator/ErrorCondenser.ts (hunk 2 above)
Comment (pmstss) [reply packages/validator/src/validator/ErrorCondenser.ts:82]: seems invalid - as it's special case w/o "initialValue" (few lines latter we have e. g. `!prev.params`)

File: packages/validator/src/validator/ErrorCondenser.ts [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+          this.detectMostFrequentMessageNames(path).map(
+            (name) => this.tree[path][name]
+          );
+
+        return res.concat(
+          frequentMessageErrors.map(this.mergeDuplicatedErrors.bind(this))
+        );
+      },
+      []
+    );
```
[the remark this answers is not in the document]
Comment (pmstss) [reply packages/validator/src/validator/ErrorCondenser.ts]: utilized `flatMap`

File: packages/validator/tests/error-condenser.spec.ts (hunk 14 above)
Comment (pmstss) [reply packages/validator/tests/error-condenser.spec.ts:14]: but they are already in `ErrorCondenser` suite; please clarify the idea of having one more nesting for suites

Comment (derevnjuk) [reply packages/validator/tests/error-condenser.spec.ts:14]: Just to have the following structure: `class -> method -> tests`. That provides a huge amount of benefits (e.g skipping all tests for particular method, removing them without of pain, easy navigation).

File: packages/validator/src/schemas/postman/v2.1.0.json (hunk 7 above)
Comment (pmstss) [reply packages/validator/src/schemas/postman/v2.1.0.json:640]: not really as "anyOf" validator messages are nearly good due to const restrictions on type for both branches; added minor fixes and tests

File: packages/validator/src/schemas/openapi/v2.0.0.json (hunk 11 above)
Comment (pmstss) [reply packages/validator/src/schemas/openapi/v2.0.0.json:792]: not sure that got your question correctly.

> type is required when in is not body.
> schema must be required if in is body (btw there is not schema def here)

this is reflected in `bodyParameter` declaration above; but this comment is linked to specific `pathParameterSubSchema`

File: packages/validator/src/schemas/openapi/v3.0.0.json (hunk 10 above)
Comment (pmstss) [reply packages/validator/src/schemas/openapi/v3.0.0.json:1128]: seems it's already covered by `allOf` on next line + inner `allOf` of `ParameterLocation`:

```
        {
          "if": {
            "required": ["in"],
            "properties": {
              "in": {
                "const": "path"
              }
            }
          },
          "then": {
            "description": "Parameter in path",
            "required": ["in", "required"],  // <-- HERE
            "properties": {
              "in": {
                "const": "path"
              },
              "style": {
                "enum": ["matrix", "label", "simple"],
                "default": "simple"
              },
              "required": {
                "const": true // <-- HERE
              }
            }
          }
        }
```

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**